### PR TITLE
xbmgmt flash to support apu_versal flash_type

### PR DESCRIPTION
**Pull request by xdavidz**
This is the code for flashing APU linux by using xbmgmt legacy command.

We possibly need to switch to xgq but this version gives us a working solution first.

### CODE REVIEW

**Comment by dayeh-xilinx** [comment]
retest this please. XRT pipeline failed when downloading docker image from artifactory. Try again.

File: src/runtime_src/core/pcie/tools/xbmgmt/flasher.h [hunk 1]
```diff
@@ -94,8 +95,9 @@ class Flasher
         BPI,
         QSPIPS,
         OSPIVERSAL,
+        APUVERSAL,
     };
-    const char *E_FlasherTypeStrings[5] = { "UNKNOWN", "SPI", "BPI", "QSPI_PS", "OSPI_VERSAL"};
+    const char *E_FlasherTypeStrings[6] = { "UNKNOWN", "SPI", "BPI", "QSPI_PS", "OSPI_VERSAL", "APU_VERSAL"};
```
[the remark this answers is not in the document]
Comment (keryell) [reply src/runtime_src/core/pcie/tools/xbmgmt/flasher.h:100]: Perhaps
```c++
constexpr char *E_FlasherTypeStrings[] = { "UNKNOWN", "SPI", "BPI", "QSPI_PS", "OSPI_VERSAL", "APU_VERSAL"};
```
?

File: src/runtime_src/core/pcie/tools/xbmgmt/xapuversal.cpp [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+XAPUVER_Flasher::XAPUVER_Flasher(std::shared_ptr<pcidev::pci_device> dev)
+{
+    mDev = dev;
+}
+
+/**
+ * @brief XQSPIPS_Flasher::~XQSPIPS_Flasher
+ */
+XAPUVER_Flasher::~XAPUVER_Flasher()
+{
```
[the remark this answers is not in the document]
Comment (keryell) [reply src/runtime_src/core/pcie/tools/xbmgmt/xapuversal.cpp:34]: But is there a need for this?

**Review by larry9523 — APPROVED** [review]
Looks good to me. This seems to be a temporary solution. Once we move to XGQ, we are not supposed to use ospi_versal sub-driver, right?

**Comment by xdavidz** [comment]
> Looks good to me. This seems to be a temporary solution. Once we move to XGQ, we are not supposed to use ospi_versal sub-driver, right?


We will not use ospi_versal driver once we switch to XGQ. This is the minimum non-harmful change to make RPU flash work.

However, is it safe to add additional "XRT_XFR_PKT_FLAGS_APU_PDI"? 
If there is concern about this flag, we can hold off the change in master XRT, and use my specific build to temporary deliverables.

In XGQ solution, I guess we will have entirely different design.

**Comment by larry9523** [comment]
> > Looks good to me. This seems to be a temporary solution. Once we move to XGQ, we are not supposed to use ospi_versal sub-driver, right?
> 
> We will not use ospi_versal driver once we switch to XGQ. This is the minimum non-harmful change to make RPU flash work.
> 
> However, is it safe to add additional "XRT_XFR_PKT_FLAGS_APU_PDI"?
> If there is concern about this flag, we can hold off the change in master XRT, and use my specific build for temporary deliverables.
> 
> In XGQ solution, I guess we will have entirely different design.

It should be fine. ospi_versal driver should stop evolving anyway. As long as it does not break any existing platform.

**Review by houlz0507 — APPROVED** [review]

**Comment by xdavidz** [comment]
> > > Looks good to me. This seems to be a temporary solution. Once we move to XGQ, we are not supposed to use ospi_versal sub-driver, right?
> > 
> > 
> > We will not use ospi_versal driver once we switch to XGQ. This is the minimum non-harmful change to make RPU flash work.
> > However, is it safe to add additional "XRT_XFR_PKT_FLAGS_APU_PDI"?
> > If there is concern about this flag, we can hold off the change in master XRT, and use my specific build for temporary deliverables.
> > In XGQ solution, I guess we will have entirely different design.
> 
> It should be fine. ospi_versal driver should stop evolving anyway. As long as it does not break any existing platform.

Thanks, I will add @AShivangi to review it and also address other code lint warnings.

**Comment by xdavidz** [comment]
closed, moving to xbmgmt2 + xgq

**Pull request closed without merging**